onnxruntime/core/session/inference_session.cc

```diff
@@ -1864,6 +1873,75 @@ AllocatorPtr InferenceSession::GetAllocator(const OrtMemoryInfo& mem_info) const
   return session_state_->GetAllocator(mem_info);
 }
 
+void InferenceSession::ShrinkDefaultMemoryArenas(const std::string& ort_device_list) {
+  if (ort_device_list.empty()) {
+    return;
+  }
+
+  std::stringstream ss_1(ort_device_list);
+  std::string device_id_pair;
+
+  // Process all device-id pair(s)
+  while (std::getline(ss_1, device_id_pair, ';')) {
+    std::stringstream ss_2(device_id_pair);
+    std::string device_id_component;
+
+    // default values
+    OrtDevice::DeviceType device_type = -1;
+    OrtDevice::MemoryType memory_type = OrtDevice::MemType::DEFAULT;
+    OrtDevice::DeviceId device_id = 0;
+
+    int iter = 0;
+    // Process this device-id pair
+    while (std::getline(ss_2, device_id_component, ':')) {
+      if (iter == 0) {  // this component corresponds to device
+        if (device_id_component == "cpu") {
+          device_type = OrtDevice::CPU;
+        } else if (device_id_component == "gpu") {
+          device_type = OrtDevice::GPU;
+        } else {
+          LOGS(*session_logger_, ERROR) << "Unsupported device specified in the memory arena shrink list: " << device_id_component
+                                        << " and shrinking is skipped for this";
+          break;
+        }
+      } else if (iter == 1) {  // This component corresponds to device id
+        if (!TryParseStringWithClassicLocale<OrtDevice::DeviceId>(device_id_component, device_id)) {
+          LOGS(*session_logger_, ERROR) << "Unsupported device id in the memory arena shrink list: " << device_id_component
+                                        << " and shrinking is skipped for this";
+          break;
+        }
+      }
+
+      ++iter;
+    }
+
+    // Shrink if it is an arena based allocator
+    auto alloc = session_state_->GetAllocator(OrtDevice(device_type, memory_type, device_id));
+
+    if (alloc == nullptr) {
+      LOGS(*session_logger_, ERROR) << "Did not find an arena based allocator registered for device-id "
```

Comment:
WARNING if you can continue and allow the user to correct the invalid input next time, ERROR if it doesn't make sense to continue and want the user to correct the problem immediately.

Reply:
This goes away if we return error status in the intiial validation